### PR TITLE
Update doc for kafe2go

### PR DESCRIPTION
**Pull request by cverstege**
@JohannesGaessler can you please proofread the changes and maybe help me with the advanced data errors for kafe2go?

### CODE REVIEW

File: doc/src/parts/mathematical_foundations.rst [hunk 1]
```diff
@@ -227,33 +240,72 @@ As it turns the method of least squares is a special case of the method
...
+penalty to the cost function.
+The value of the penalty is determined by how well the parameter lies within its constraints.
+The new cost function then reads as
 
-:math:`\chi^2`
+.. math::
 
-Negative log-likelihood estimator
----------------------------------
+    C_\mathrm{total} = C + \mathrm{constraints}
```
Comment (JohannesGaessler) [inline doc/src/parts/mathematical_foundations.rst:298]: \mathrm{constraints} seems like a weird choice for a variable name in LaTeX.
Is this intentional?

File: doc/src/parts/mathematical_foundations.rst [hunk 2]
```diff
@@ -227,33 +240,72 @@ As it turns the method of least squares is a special case of the method
...
 
+In general the penalty is given by the current parameter values :math:`\textbf{p}`, the expected
+parameter values :math:`\bm{\mu}` and the covariance matrix :math:`V`:
```
Comment (JohannesGaessler) [inline doc/src/parts/mathematical_foundations.rst:301]: This makes it sound like it's the same covariance matrix as used for the regular cost function.

**Review by JohannesGaessler — CHANGES_REQUESTED** [review]
Partial review: devloper_guide.rst and mathematical_foundations.rst

File: doc/src/parts/mathematical_foundations.rst (hunk 1 above)
Comment (cverstege) [reply doc/src/parts/mathematical_foundations.rst:298]: I didn't know how to call it, so I just wrote constaints. Do you have any ideas?

Comment (JohannesGaessler) [reply doc/src/parts/mathematical_foundations.rst:298]: I would just use indices for C.
So something like `C_total = C_0 + C_constraint` with some previous hint that `C \equiv C_0`

File: doc/src/parts/mathematical_foundations.rst [hunk 3]
```diff
@@ -50,11 +47,25 @@ correlated errors. In order to make meaningful quantitative
 statements based on measurement data it is essential to quantify
 both components of all uncertainties, expressed in terms of
 their covariance matrix, or by the uncertainties and
-their correlation matrix. 
+their correlation matrix.
```
[the remark this answers is not in the document]
Comment (cverstege) [reply doc/src/parts/mathematical_foundations.rst:50]: It says that systematical and statistical uncertainties can be described through independent and correlated erros. Do you have an idea on how to change this/ rewrite that section?

File: doc/src/parts/mathematical_foundations.rst [hunk 4]
```diff
@@ -227,33 +240,72 @@ As it turns the method of least squares is a special case of the method
...
+Depending on the chosen cost function, uncertainties are handled differently.
+Simple pointwise gaussian uncertainties can be handled by a negative log likelihood cost function
+with a gaussian distribution for each data point.
+Alternatively the method of least squares can be used which can handle correlations between the
+uncertainties via the covariance or correlation matrix as well.
+The squared pointwise uncertainties correspond to the diagonal elements of the covariance matrix.
 
-Handling uncertainties
-======================
+Uncertainties relative to the data are constant, as the data is constant when performing a fit.
```
[the remark this answers is not in the document]
Comment (cverstege) [reply doc/src/parts/mathematical_foundations.rst:253]: How uncertainties work and are handled is imo part of the theoretical foundations. There currently is no other place where we can put this. The User Guide will only contain examples on how to use it, but won't explain how it works in the back end.

[the remark this answers is not in the document]
Comment (cverstege) [reply doc/src/parts/mathematical_foundations.rst:253]: I thought you meant the whole section. Only the part which cost functions can handle which uncertainties can indeed be moved.

**Comment by cverstege** [comment]
@JohannesGaessler I just pushed a new, cleaned up and finished version. It should include all your earlier comments.
After a review of each changed file and fixing I would merge those commits.

File: doc/src/parts/user_guide_kafe2go.rst [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+.. code-block:: yaml
+
+    fixed_parameters:
+      a: 1
+      b: 11.5
+    limited_parameters:
+      c: [0.0, 1.0]
+      d: [-5, 5]
+
+Note, that a limited parameter needs a lower and an upper limit. The first value of the list is the
```
Comment (JohannesGaessler) [inline doc/src/parts/user_guide_kafe2go.rst:323]:
```suggestion
Note that a limited parameter needs a lower and an upper limit. The first value of the list is the
```

File: doc/src/parts/user_guide_kafe2go.rst [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+lower limit and the second value is the upper limit.
+If a parameter shall be limited to only positive values, choose 0 as lower and a high value as
```
Comment (JohannesGaessler) [inline doc/src/parts/user_guide_kafe2go.rst:325]:
```suggestion
If a parameter shall be limited to non-negative values, choose 0 as lower and a high value as
```

**Review by JohannesGaessler — CHANGES_REQUESTED** [review]
I think you misunderstood what I wanted you to write regarding statistical/systematic errors.
I will push a commit with text that I think would be appropriate.

File: doc/src/parts/developer_guide.rst [hunk 7]
```diff
@@ -90,3 +90,32 @@ To build the documentation run::
...
+`Sphinx docstring format <https://sphinx-rtd-tutorial.readthedocs.io/en/latest/docstrings.html>`_.
+
+Since we also support Python 2, we can't use Python's own `typehints <https://www.python
+.org/dev/peps/pep-0484/>`_ introduced with Python 3.5.
+Modern IDEs can also perform type checking with the help of docstrings.
+That's why, in general, we try to include the types inside the docstrings.
+
+If you're working on the code and find missing, incomplete or hard do understand docstrings,
+please update them and create a `Pull Request <https://github.com/dsavoiu/kafe2/compare>`_ or
+`Issue <https://github.com/dsavoiu/kafe2/issues/new/choose>`_ on GitHub.
```
[the remark this answers is not in the document]
Comment (cverstege) [reply doc/src/parts/developer_guide.rst:121]: Removed them.

File: doc/src/parts/user_guide_kafe2go.rst [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+When using kafe2go no programming is required.
+Instead all necessary information for performing a fit is defined inside a so called
+`YAML <https://en.wikipedia.org/wiki/YAML>`_ file.
+Full examples can be found in the :ref:`beginners_guide`.
+
+To run kafe2go issue this command in the terminal, after installing *kafe2*:
+
+.. code-block:: bash
+
+    kafe2go path/to/fit.yml
```
[the remark this answers is not in the document]
Comment (cverstege) [reply doc/src/parts/user_guide_kafe2go.rst:20]: Maybe we should add a `kafe2go.bat` file if that helps?

File: doc/src/parts/user_guide.rst [hunk 9]
```diff
@@ -22,11 +22,296 @@ correlations.
 .. figure:: ../_static/img/kafe2_structure.png
     :alt: General workflow with kafe2.
 
+
+Datasets
+========
+When performing fits with *kafe2*, the data is stored in so-called data containers
+(:py:obj:`~.DataContainerBase`-derived objects).
+The difference between the container classes comes down the type of data they store.
```
Comment (JohannesGaessler) [inline doc/src/parts/user_guide.rst:30]:
```suggestion
The difference between the container classes comes down to the type of data they store.
```

File: doc/src/parts/user_guide.rst [hunk 10]
```diff
@@ -22,11 +22,296 @@ correlations.
...
+* One-dimensional data can either be kept as-is or filled into a histogram ("*binning*").
+  In *kafe2* these types of data and their corresponding fits are referred to as unbinned and
+  histogram data/fits
+  (:py:obj:`~.UnbinnedContainer` and :py:obj:`~.HistContainer`).
+* Two-dimensional data and the corresponding fit is referred to as XY data/fit
+  (:py:obj:`~.XYContainer`).
+
+Setting the data
+----------------
+Data creators are created as regular Python objects from iterables (lists, arrays, etc.) of floats.
```
Comment (JohannesGaessler) [inline doc/src/parts/user_guide.rst:48]:
```suggestion
Data containers are created as regular Python objects from iterables (lists, arrays, etc.) of floats.
```

File: doc/src/parts/user_guide.rst [hunk 11]
```diff
@@ -22,11 +22,296 @@ correlations.
...
+    # By default, a linear function f=a*x+b will be used as the model function.
+    line_fit = Fit(data=xy_data)
+
+Alternatively :py:obj:`~.XYFit`, :py:obj:`~.HistFit`, :py:obj:`~.UnbinnedFit` or
+:py:obj:`~.IndexedFit` can be used to create fits with corresponding datasets.
+
+Setting a model function
+------------------------
+
+*kafe2* fit objects accept normal python functions as model functions.
```
Comment (JohannesGaessler) [inline doc/src/parts/user_guide.rst:209]:
```suggestion
*kafe2* fit objects accept normal Python functions as model functions.
```

File: doc/src/parts/user_guide_kafe2go.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (JohannesGaessler) [reply doc/src/parts/user_guide_kafe2go.rst:20]: That is not the problem.
The problem is that Windows doesn't support shebangs.
On Windows the file extension determines how a file is interpreted.

File: doc/src/parts/mathematical_foundations.rst [hunk 12]
```diff
@@ -227,33 +221,65 @@ As it turns the method of least squares is a special case of the method
...
+
+As the variance of a Poisson distribution is equal to the mean value, no uncertainties have to
+be given when using Poisson uncertainties.
+
+Parameter constraints
+---------------------
+When performing a fit, some values of the model function might have already been determined in
+previous experiments.
+Those results and uncertainties can then be used to constrain the given parameters in a new fit.
+This eradicates the need to manually propagate the uncertainties on the final fit results, as
```
Comment (JohannesGaessler) [inline doc/src/parts/mathematical_foundations.rst:261]:
```suggestion
This eliminates the need to manually propagate the uncertainties on the final fit results, as
```

File: doc/src/parts/user_guide.rst [hunk 13]
```diff
@@ -22,11 +22,296 @@ correlations.
...
+
+.. _constraints_guide:
+
+Parameter Constraints
+---------------------
+
+When performing a fit, some values of the model function might have already been determined in
+previous experiments.
+Those results and uncertainties can then be used to constrain the given parameters in a new fit.
+This eradicates the need to manually propagate the uncertainties on the final fit results, as
```
Comment (JohannesGaessler) [inline doc/src/parts/user_guide.rst:267]:
```suggestion
This eliminates the need to manually propagate the uncertainties on the final fit results, as
```

**Review by JohannesGaessler — APPROVED** [review]
This is basically good to merge now; the only problem were a few typos/grammatical mistakes half of which were actually introduced by my own suggestions.

**Comment by cverstege** [comment]
Fixed all known issues, including line breaks for long formulas and replaced `\textbf` with `\bm` in formulas for "correct" boldness in math mode.
Going to rebase and merge after next force push.

**Pull request merged**